xds/src/main/java/io/grpc/xds/XdsRoutingLoadBalancerProvider.java

```diff
@@ -0,0 +1,370 @@
+              "No child policy for action " + name + " in xds_routing LB policy:\n "
+                  + rawConfig));
+        }
+        boolean targetParsingSucceeded = false;
+        for (LbConfig lbConfig : childConfigCandidates) {
+          String policyName = lbConfig.getPolicyName();
+          LoadBalancerProvider lbProvider = loadBalancerRegistry().getProvider(policyName);
+          if (lbProvider == null) {
+            logger.log(
+                Level.FINEST,
+                "The policy {0} for is not available in xds_routing LB policy:\n {1}",
+                new Object[]{policyName, name});
+          } else {
+            ConfigOrError parsedLbPolicyConfig = lbProvider
+                .parseLoadBalancingPolicyConfig(lbConfig.getRawConfigValue());
+            if (parsedLbPolicyConfig.getError() != null) {
+              // Based on service config error-handling spec, if the chosen config is found invalid
+              // while other configs that come later were valid, the gRPC config would still be
+              // considered invalid as a whole.
+              return parsedLbPolicyConfig;
+            }
+            parsedActions.put(
+                name,
+                new ChildConfig(policyName, parsedLbPolicyConfig.getConfig()));
+            targetParsingSucceeded = true;
+            break;
+          }
+        }
+        if (!targetParsingSucceeded) {
+          return ConfigOrError.fromError(Status.INTERNAL.withDescription(
+              "No child policy available for action " + name + " in xds_routing LB policy:\n "
+                  + rawConfig));
+        }
+      }
+
+      List<Map<String, ?>> routes = JsonUtil.getListOfObjects(rawConfig, "route");
+      if (routes == null || routes.isEmpty()) {
+        return ConfigOrError.fromError(Status.INTERNAL.withDescription(
+            "No routes provided for xds_routing LB policy:\n" + rawConfig));
+      }
+      List<Route> parsedRoutes = new ArrayList<>();
+      Set<MethodName> methodNames = new HashSet<>();
+      for (int i = 0; i < routes.size(); i++) {
+        Map<String, ?> route = routes.get(i);
+        String actionName = JsonUtil.getString(route, "action");
+        if (actionName == null) {
+          return ConfigOrError.fromError(Status.INTERNAL.withDescription(
+              "No action name provided for one of the routes in xds_routing LB policy:\n"
+                  + rawConfig));
+        }
+
+        if (!parsedActions.containsKey(actionName)) {
+          return ConfigOrError.fromError(Status.INTERNAL.withDescription(
+              "No action defined for route " + route + " in xds_routing LB policy:\n" + rawConfig));
+        }
+
+        Map<String, ?> methodName = JsonUtil.getObject(route, "methodName");
+        if (methodName == null) {
+          return ConfigOrError.fromError(Status.INTERNAL.withDescription(
+              "No methodName provided for one of the routes in xds_routing LB policy:\n"
```

Comment:
s/methodName/method_name/g

Error message for input data, not implementation details.